File: backend/app/schemas/product.py
from pydantic import BaseModel
from typing import List, Optional
from decimal import Decimal
from enum import Enum

class DependencyType(str, Enum):
    requires = "requires"
    excludes = "excludes"

# Base schemas
class PartOptionBase(BaseModel):
    name: str
    base_price: Decimal
    in_stock: bool = True

class PartTypeBase(BaseModel):
    name: str

class ProductBase(BaseModel):
    name: str
    category: str
    is_active: bool = True
    featured: bool = False
    base_price: Optional[Decimal] = None
    image_url: Optional[str] = None

class OptionDependencyBase(BaseModel):
    depends_on_option_id: int
    type: DependencyType

class ConditionalPriceBase(BaseModel):
    condition_option_id: int
    conditional_price: Decimal

# Create schemas
class PartOptionCreate(PartOptionBase):
    pass

class PartTypeCreate(PartTypeBase):
    pass

class ProductCreate(ProductBase):
    pass

class OptionDependencyCreate(OptionDependencyBase):
    pass

class ConditionalPriceCreate(ConditionalPriceBase):
    pass

# Read schemas
class PartOption(PartOptionBase):
    id: int
    part_type_id: int

    class Config:
        orm_mode = True

class PartType(PartTypeBase):
    id: int
    product_id: int
    options: List[PartOption] = []

    class Config:
        orm_mode = True

class Product(ProductBase):
    id: int
    part_types: List[PartType] = []

    class Config:
        orm_mode = True

class OptionDependency(OptionDependencyBase):
    id: int
    option_id: int

    class Config:
        orm_mode = True

class ConditionalPrice(ConditionalPriceBase):
    id: int
    option_id: int

    class Config:
        orm_mode = True

# Full detailed schemas with relationships
class PartOptionDetail(PartOption):
    dependencies: List[OptionDependency] = []
    conditional_prices: List[ConditionalPrice] = []

class PartTypeDetail(PartType):
    options: List[PartOptionDetail] = []

class ProductDetail(Product):
    part_types: List[PartTypeDetail] = [] 

File: frontend/src/types/product.ts
export enum DependencyType {
  REQUIRES = "requires",
  EXCLUDES = "excludes"
}

export interface OptionDependency {
  id: number;
  option_id: number;
  depends_on_option_id: number;
  type: DependencyType;
}

export interface ConditionalPrice {
  id: number;
  option_id: number;
  condition_option_id: number;
  conditional_price: number;
}

export interface PartOption {
  id: number;
  name: string;
  base_price: number;
  in_stock?: boolean;
  part_type_id?: number;
  is_compatible?: boolean;
  available_for_selection?: boolean;
  availability_reason?: string;
  dependencies?: OptionDependency[];
  conditional_prices?: ConditionalPrice[];
}

export interface PartType {
  id: number;
  name: string;
  product_id?: number;
  options: PartOption[];
}

export interface Product {
  id: number;
  name: string;
  category: string;
  description?: string;
  basePrice: number;
  image: string;
  image_url?: string;
  is_active?: boolean;
  featured?: boolean;
  partTypes: PartType[];
}

// Interface for available options for a specific product
export interface AvailableOption {
  id: number;
  name: string;
  base_price: number;
  is_compatible: boolean;
  available_for_selection?: boolean;
  selected?: boolean;
  availability_reason?: string;
}

export interface AvailablePartType {
  id: number;
  name: string;
  options: AvailableOption[];
}

// Interface for add to cart request
export interface AddToCartRequest {
  product_id: number;
  selected_options: number[];
  quantity: number;
} 